lib/os/queued_operation.c

```diff
@@ -0,0 +1,531 @@
+		goto out;
+	}
+
+	k_spinlock_key_t key = k_spin_lock(&mgr->lock);
+	sys_slist_t *list = &mgr->operations;
+	u32_t state = mgr->state;
+
+	/* Preserve error state, or insert the item into the list. */
+	if (state_is_error(state)) {
+		rv = -ENODEV;
+	} else if (priority == QUEUED_OPERATION_PRIORITY_PREPEND) {
+		sys_slist_prepend(list, &op->node);
+	} else if (priority == QUEUED_OPERATION_PRIORITY_APPEND) {
+		sys_slist_append(list, &op->node);
+	} else {
+		struct queued_operation *prev = NULL;
+		struct queued_operation *tmp;
+
+		SYS_SLIST_FOR_EACH_CONTAINER(list, tmp, node) {
+			if (priority < op_get_priority(tmp)) {
+				break;
+			}
+			prev = tmp;
+		}
+
+		if (prev == NULL) {
+			sys_slist_prepend(list, &op->node);
+		} else {
+			sys_slist_insert(list, &prev->node, &op->node);
+		}
+	}
+
+	/* Initiate an operation only if we're off. */
+	if (state == ST_OFF) {
+		start_and_unlock(mgr, key);
+	} else {
+		__ASSERT_NO_MSG(state != ST_IDLE);
+		k_spin_unlock(&mgr->lock, key);
+	}
+
+out:
+	/* Preserve a service-specific success code on success */
+	if ((rv >= 0) && (validate_rv >= 0)) {
+		rv = validate_rv;
+	}
+
+	return rv;
+}
+
+void queued_operation_finalize(struct queued_operation_manager *mgr,
+			       int res)
+{
+	__ASSERT_NO_MSG(mgr != NULL);
+
+	k_spinlock_key_t key = k_spin_lock(&mgr->lock);
+	struct queued_operation *op = mgr->current;
+	u32_t state = mgr->state;
+	bool processing = (state == ST_PROCESSING);
+
+	__ASSERT_NO_MSG(op != NULL);
+	__ASSERT_NO_MSG((state == ST_NOTIFYING)
+			|| (state == ST_PROCESSING));
+
+	mgr->state = ST_FINALIZING;
+
+	k_spin_unlock(&mgr->lock, key);
+
+	finalize_and_notify(mgr, op, res);
+
+	/* If we were processing we need to reselect; if we were
+	 * notifying we'll reselect when the notification completes.
+	 */
+	if (processing) {
+		select_next_and_unlock(mgr, k_spin_lock(&mgr->lock));
+	}
+}
+
+int queued_operation_cancel(struct queued_operation_manager *mgr,
+			    struct queued_operation *op)
+{
+	__ASSERT_NO_MSG(mgr != NULL);
+	__ASSERT_NO_MSG(op != NULL);
+
+	int rv = 0;
+	k_spinlock_key_t key = k_spin_lock(&mgr->lock);
+
+	if (op == mgr->current) {
+		rv = -EINPROGRESS;
+	} else if (!sys_slist_find_and_remove(&mgr->operations, &op->node)) {
+		rv = -EINVAL;
+	}
+
+	k_spin_unlock(&mgr->lock, key);
+
+	if (rv == 0) {
+		finalize_and_notify(mgr, op, -ECANCELED);
+	}
+
+	return rv;
+}
+
+bool queued_operation_has_error(struct queued_operation_manager *mgr)
+{
+	__ASSERT_NO_MSG(mgr != NULL);
+
+	k_spinlock_key_t key = k_spin_lock(&mgr->lock);
+
+	bool rv = state_is_error(mgr->state);
+
+	k_spin_unlock(&mgr->lock, key);
+
+	return rv;
+}
+
+static void reset_callback(struct onoff_manager *oomgr,
+			   struct onoff_client *cli,
+			   u32_t state,
+			   int res)
+{
+	struct queued_operation_manager *mgr
+		= CONTAINER_OF(cli, struct queued_operation_manager, onoff_client);
+
+	settle_onoff(mgr, ST_ERROR, res);
+}
+
+int queued_operation_reset(struct queued_operation_manager *mgr,
```

Comment:
if onoff would be moved from the manager then reset function would just set it to off state after successful external reset. Imo, that would simplify queue operation manager since it would decouple resetting service from operations queueing.

Reply:
Per above I'm keeping onoff in the manager because it handles the edge cases during transitions.